storage/innobase/fil/fil0fil.cc

```diff
@@ -539,7 +539,7 @@ bool fil_node_t::read_page0(bool first)
 	/* Align the memory for file i/o if we might have O_DIRECT set */
 	byte* page = static_cast<byte*>(ut_align(buf2, psize));
 	IORequest request(IORequest::READ);
-	if (!os_file_read(request, handle, page, 0, psize)) {
+	if (os_file_read(request, handle, page, 0, psize) != DB_SUCCESS) {
```

Comment:
It turns out that the return type of `os_file_read_func()` was changed in https://github.com/mysql/mysql-server/commit/98909cefbc37e54efc6452c7e95bccbf64ac9213 (MySQL 5.7) from `ibool` to `dberr_t`. The reviewer (if there was any) failed to point out that because of future merges, it could be a bad idea to change the return type of a function without changing the function name.

This change was applied to MariaDB 10.2.2 in commit 2e814d4702d71a04388386a9f591d14a35980bfe, but the MariaDB-specific function (then in `fil_node_open_file()`) was not fully adjusted accordingly in the same commit. Essentially, the `if` statement body became dead code in MariaDB, and I suppose that we would be dealing with an uninitialized buffer after a failed page read.

Can you try to write a test case for this, and extend the MDEV-19231 ticket as well as this fix with the details? I hope that it will suffice to play with `chmod`. There might be a `stat()` call that would prevent InnoDB from opening directories or special files (such as when you used `mkdir` or `mknod` to create an `.ibd` file). You might use add `DBUG_EXECUTE_IF` to bypass some checks so that the execution will actually proceed this far.

For manual testing, you might create an error-injected block device with `dmsetup`; see https://github.com/ringerc/scrapcode/blob/master/testcases/fsync-error-clear/standalone/fsync-error-clear.c for details. That obviously can’t be deployed in contiguous integration systems.

Reply:
`chmod -r ` didn't help me because it make file unreadable and it's even not opened.

```diff
diff --git a/storage/innobase/fil/fil0fil.cc b/storage/innobase/fil/fil0fil.cc
index ccf1ac0ecef..b80a491f01f 100644
--- a/storage/innobase/fil/fil0fil.cc
+++ b/storage/innobase/fil/fil0fil.cc
@@ -539,7 +539,9 @@ bool fil_node_t::read_page0(bool first)
        /* Align the memory for file i/o if we might have O_DIRECT set */
        byte* page = static_cast<byte*>(ut_align(buf2, psize));
        IORequest request(IORequest::READ);
-       if (os_file_read(request, handle, page, 0, psize) != DB_SUCCESS) {
+       dberr_t err = os_file_read(request, handle, page, 0, psize);
+       DBUG_EXECUTE_IF("fil_node_t_read_page0_failure", err = DB_ERROR;);
+       if (err != DB_SUCCESS) {
                ib::error() << "Unable to read first page of file " << name;
                ut_free(buf2);
                return false;
```
This didn't work because `fil_node_t::read_page0()` is called to early at the startup.

I've tried this for the manual testing.
```diff
+++ b/storage/innobase/fil/fil0fil.cc
@@ -539,7 +539,14 @@ bool fil_node_t::read_page0(bool first)
        /* Align the memory for file i/o if we might have O_DIRECT set */
        byte* page = static_cast<byte*>(ut_align(buf2, psize));
        IORequest request(IORequest::READ);
-       if (os_file_read(request, handle, page, 0, psize) != DB_SUCCESS) {
+       dberr_t err = os_file_read(request, handle, page, 0, psize);
+       static int skip = 0;
+       if (skip > 0) {
+               skip--;
+       } else {
+               err = DB_ERROR;
+       }
+       if (err != DB_SUCCESS) {
                ib::error() << "Unable to read first page of file " << name;
                ut_free(buf2);
                return false;
```

Depending on `skip` value I saw different failures:
`skip = 0`:
```c++
#0  __pthread_kill (threadid=<optimized out>, signo=6) at pthread_kill.c:56
#1  0x00000000012869db in my_write_core (sig=6) at stacktrace.c:477
#2  0x00000000009a2ca9 in handle_fatal_signal (sig=6) at signal_handler.cc:305
#3  <signal handler called>
#4  __GI_raise (sig=sig@entry=6) at raise.c:50
#5  0x00007f6bd18bf535 in __GI_abort () at abort.c:79
#6  0x0000000000ec8340 in ut_dbg_assertion_failed (expr=0x143797a "0", file=0x14c7232 "/work/mariadb/storage/innobase/fil/fil0fil.cc", line=1856) at ut0dbg.cc:60
#7  0x0000000000fde6f5 in fil_open_log_and_system_tablespace_files () at fil0fil.cc:1856
#8  0x0000000000e3ba33 in innobase_start_or_create_for_mysql () at srv0start.cc:2077
#9  0x0000000000c22d1d in innobase_init (p=0x28a50d8) at ha_innodb.cc:4377
#10 0x00000000009a419c in ha_initialize_handlerton (plugin=0x2874db8) at handler.cc:521
#11 0x000000000070c195 in plugin_initialize (tmp_root=0x7ffcfa892fa0, plugin=0x2874db8, argc=0x1b9f2b0 <remaining_argc>, argv=0x2730698, options_only=false) at sql_plugin.cc:1416
#12 0x000000000070bb04 in plugin_init (argc=0x1b9f2b0 <remaining_argc>, argv=0x2730698, flags=2) at sql_plugin.cc:1697
#13 0x00000000005e17ef in init_server_components () at mysqld.cc:5311
#14 0x00000000005de200 in mysqld_main (argc=149, argv=0x2730698) at mysqld.cc:5907
#15 0x00000000005d9f32 in main (argc=26, argv=0x7ffcfa8937b8) at main.cc:25
```

`skip = 1`:
```c++
create table t (a int) engine=innodb;
insert into t values (1);
drop table t;
innodb.kevg 'innodb'                     [ fail ]  Found warnings/errors in server log file!
        Test ended at 2019-04-25 10:02:06
line
2019-04-25 10:02:06 140467501000448 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_index_stats.ibd
2019-04-25 10:02:06 140467501000448 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_index_stats.ibd
2019-04-25 10:02:06 140467501000448 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:06 140467501000448 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_index_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_index_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_index_stats.ibd
2019-04-25 10:02:06 140467492607744 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_index_stats.ibd
2019-04-25 10:02:06 140468052047616 [ERROR] InnoDB: Table `mysql`.`innodb_table_stats`is corrupted
^ Found warnings in /home/kevg/work/debug/mysql-test/var/log/mysqld.1.err
o
```

`skip = 2`:
```c++
create table t (a int) engine=innodb;
insert into t values (1);
drop table t;
innodb.kevg 'innodb'                     [ fail ]  Found warnings/errors in server log file!
        Test ended at 2019-04-25 10:02:46
line
2019-04-25 10:02:46 140686972131072 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:46 140686972131072 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:46 140686552725248 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:46 140686552725248 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:46 140686552725248 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:46 140686552725248 [ERROR] InnoDB: Unable to read first page of file ./mysql/innodb_table_stats.ibd
2019-04-25 10:02:46 140687389886208 [ERROR] InnoDB: Table `mysql`.`innodb_table_stats`is corrupted
^ Found warnings in /home/kevg/work/debug/mysql-test/var/log/mysqld.1.err
```

`skip > 2` works fine.

So, server refuses to start because of that error which seems reasonable. No trying to work in incorrect state.

Reply:
Thanks!